diesel/src/sqlite/connection/stmt.rs

```diff
@@ -189,88 +188,113 @@ impl Drop for Statement {
-                    binds_to_free.push((bind_idx, None));
-                }
-                SqliteBindValue::Binary(b) => {
-                    binds_to_free.push((bind_idx, Some(SqliteBindValue::Binary(b))));
-                }
-                SqliteBindValue::String(b) => {
-                    binds_to_free.push((bind_idx, Some(SqliteBindValue::String(b))));
-                }
-                _ => (),
+        // This does not need to change based on the query. For compilation performance and binary
+        // size, this will be a non-generic function within Diesel.
+        non_generic_inner(&mut bound_statement, bind_collector)?;
+        fn non_generic_inner(
+            bound_statement: &mut BoundStatement<'_, '_>,
+            bind_collector: SqliteBindCollector<'_>,
+        ) -> QueryResult<()> {
+            let SqliteBindCollector { binds } = bind_collector;
+            // It is useful to preallocate `binds_to_free` because it
+            // - Guarantees that pushing inside it cannot panic, which guarantees the `Drop`
+            //   impl of `BoundStatement` will always re-`bind` as needed
+            // - Avoids reallocations
+            let mut binds_to_free = Vec::with_capacity(
+                binds
+                    .iter()
+                    .filter(|&(b, _)| {
+                        matches!(
+                            b,
+                            SqliteBindValue::BorrowedString(_)
+                                | SqliteBindValue::BorrowedBinary(_)
+                                | SqliteBindValue::Binary(_)
+                                | SqliteBindValue::String(_)
+                        )
+                    })
+                    .count(),
+            );
+
+            for (bind_idx, (bind, tpe)) in (1..).zip(binds) {
+                // It's safe to call bind here as:
+                // * The type and value matches
+                // * We ensure that corresponding buffers lives long enough below
+                // * The statement is not used yet by `step` or anything else
+                unsafe { bound_statement.statement.bind(tpe, &bind, bind_idx) }?;
+
+                // The bind statement having succeeded, we should make sure to always unbind
+                match bind {
+                    SqliteBindValue::BorrowedString(_) | SqliteBindValue::BorrowedBinary(_) => {
+                        binds_to_free.push((bind_idx, None));
+                    }
+                    b @ SqliteBindValue::Binary(_) | b @ SqliteBindValue::String(_) => {
+                        binds_to_free.push((bind_idx, Some(b)));
+                    }
+                    SqliteBindValue::SmallInt(_)
+                    | SqliteBindValue::Integer(_)
+                    | SqliteBindValue::BigInt(_)
+                    | SqliteBindValue::Float(_)
+                    | SqliteBindValue::Double(_)
+                    | SqliteBindValue::Null => {}
```

Comment:
I favor exhaustive writing in this scenario as this guarantees we haven't forgotten anything.